internal/controller/apiextensions/composite/composed.go

```diff
@@ -238,3 +278,324 @@ func IsReady(_ context.Context, cd resource.Composed, t v1.ComposedTemplate) (bo
+
+	return errors.New(errMixed)
+}
+
+// RejectDuplicateNames validates that all template names are unique within the
+// supplied Composition.
+func RejectDuplicateNames(comp *v1.Composition) error {
+	seen := map[string]bool{}
+	for _, tmpl := range comp.Spec.Resources {
+		if tmpl.TemplateName == nil {
+			continue
+		}
+		if seen[*tmpl.TemplateName] {
+			return errors.New(errDuplicate)
+		}
+		seen[*tmpl.TemplateName] = true
+	}
+	return nil
+}
+
+type composedResource struct {
+	Renderer
+	ConnectionDetailsFetcher
+	ReadinessChecker
+}
+
+// An AnonymousComposer composes resources using a Composition with anonymous
+// (i.e. unnamed) resource templates. It relates composed resources to their
+// templates using their index within the composite resource's resource ref
+// array and within the composition's resource template array. It assumes the
+// length and order of the Composition is immutable; it does not garbage collect
+// resources when a template is deleted, and may leak resources when the array
+// of templates is reordered.
+type AnonymousComposer struct {
+	client   resource.ClientApplicator
+	composed composedResource
+}
+
+// Compose resources into the supplied composite per the supplied composition.
+func (r *AnonymousComposer) Compose(ctx context.Context, cr resource.Composite, comp *v1.Composition) (CompositionResult, error) {
+	refs := make([]corev1.ObjectReference, len(comp.Spec.Resources))
+	copy(refs, cr.GetResourceReferences())
+
+	cds := make([]*composed.Unstructured, len(refs))
+	for i := range refs {
+		cd := composed.New(composed.FromReference(refs[i]))
+		if err := r.composed.Render(ctx, cr, cd, comp.Spec.Resources[i]); err != nil {
+			return CompositionResult{}, errors.Wrapf(err, errFmtRenderIdx, i)
+		}
+
+		cds[i] = cd
+		refs[i] = *meta.ReferenceTo(cd, cd.GetObjectKind().GroupVersionKind())
+	}
+
+	cr.SetResourceReferences(refs)
+	if err := r.client.Update(ctx, cr); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errUpdate)
+	}
+
+	conn := managed.ConnectionDetails{}
+	ready := 0
+	for i, cd := range cds {
+		if err := r.client.Apply(ctx, cd, resource.MustBeControllableBy(cr.GetUID())); err != nil {
+			return CompositionResult{}, errors.Wrap(err, errApplyComposed)
+		}
+
+		// Connection details are fetched in all cases in a best-effort mode,
+		// i.e. it doesn't return error if the secret does not exist or the
+		// resource does not publish a secret at all.
+		c, err := r.composed.FetchConnectionDetails(ctx, cd, comp.Spec.Resources[i])
+		if err != nil {
+			return CompositionResult{}, errors.Wrap(err, errFetchSecret)
+		}
+
+		for key, val := range c {
+			conn[key] = val
+		}
+
+		rdy, err := r.composed.IsReady(ctx, cd, comp.Spec.Resources[i])
+		if err != nil {
+			return CompositionResult{}, errors.Wrap(err, errReadiness)
+		}
+
+		if rdy {
+			ready++
+		}
+	}
+
+	return CompositionResult{ConnectionDetails: conn, DesiredResources: len(comp.Spec.Resources), ReadyResources: ready}, nil
+}
```

Comment:
Note that this logic is pretty much directly extracted from the current Reconcile loop.